### DISCUSSION TITLE
Docker / container environment

### DISCUSSION THREAD

**Discussion by Ontopic** [geekjr/quickai#44] [General]
Lovely package! Not getting the traction is deserves though (yet!).

I would love to run this in a container (preferably Docker). I've less experience myself with Python's dev-tooling, so containerising "Tensorflow, PyTorch, Sklearn, Matplotlib, Numpy, and Hugging Face Transformers. You should install TensorFlow and PyTorch following the instructions from their respective websites" seems like a daunting task.

Should be easy to do though with some knowledge of Docker, Python and ai-framework-landscape brought together. From there extensions to API's, UI's and Apps can be implemented Quick as well ;)

I'm thinking a dressed down / pre-fab Dockerfile with https://github.com/ufoym/deepo as a starting point could do the trick, but interested in hearing what geekjr and others may have tried.

Thanks again!

_p.s. Please do more videos. I never was more convinced of giving a repo a try by a video than yours. No hype talk, no fake simple, no pretend awesomest new patented blabla. Nah, just the clear truth. Must be in your timing. Props_

**Comment by geekjr** [comment]
Hello, 
I will look into this further. I think being able to run QuickAI in a Docker container will improve the setup process for a lot of people.

**Comment by geekjr** [comment]
Docker support is now added! Thanks @Ontopic

**Reply by Ontopic** [reply]
Thank you! Really helpful to get people to give it a go without having the need to go through the entire setup, defeats the purpose of how quickly it allows you to test training ;)

Maybe next video with Docker? Really a start from scratch ;)

I might start a separate discussion for this, but is there a "plan" for models in the future? F.e. GPT-J, which is now ready for HuggingFace

**Reply by Ontopic** [reply]
Guess you are, just saw your commit history. Hit me up if there's any way I can assist.

**Reply by geekjr** [reply]
GPTJ is something that I have tried to add before. The only issue is that the gpu I have(2060 Super) can't run it without OOM, which means that I won't be able to test it before release.

**Comment by geekjr** [comment]
I added a GPTJ function in the main branch,

**Reply by Ontopic** [reply]
Wooo! Gonna run it right now :)

**Reply by Ontopic** [reply]
Saw on Discord btw:

On saving a usable GPT-J model after finetuning with HF:
```
def save_ckpt(model,save_dir):
    try: os.mkdir(save_dir)
    except: pass
    checkpoint = {}
    
    num_layers = len(model.state_dict())
    for i, x in tqdm(enumerate(model.state_dict().items()), total=num_layers):
      checkpoint[x[0]] = f"{save_dir}/b{i}.pt"
      params = x[1].data.clone().detach().half()
      torch.save(params, save_dir + f"/b{i}.pt")
    torch.save(checkpoint, f"{save_dir}/m.pt") 
    with open(f"{save_dir}/summary.json", 'w', encoding='utf-8') as f:
        json.dump(checkpoint, f,indent=4)
```

Not sure if useful or not, but thought might be interesting.

Where could I best contact you directly by the way for a chat?

**Reply by geekjr** [reply]
That function is very useful as it would be needed for GPTJ fine tuning support. However, due to the VRAM limitation, I would be unable to test the code, which is why I am not working on it(I was able to test the new GPTJ function in Colab on one of the free 16GB VRAM GPUS, but I don't think that is not enough for fine tuning).

**Reply by Ontopic** [reply]
I was just looking at vast.ai myself. You on any discord channel?

**Reply by Ontopic** [reply]
Also see https://github.com/huggingface/transformers/pull/13022, I think transforms is still in-flux for the upcoming hours, then probably should be easier.

**Reply by geekjr** [reply]
I see that PR is adding support of GPTJ. I mean even after that, i think ill still need more VRAM for training.

**Reply by geekjr** [reply]
Vast ai seems good rates, but I would still rather not use it.
